src/data/sonata_data.cpp

```diff
@@ -262,36 +258,50 @@ void SonataData::write_report_header() {
     hdf5_writer_->configure_attribute(reports_population_group + "/data", "units", report_units_);
 
     std::vector<uint64_t> sonata_node_ids(node_ids_);
-    convert_gids_to_sonata(sonata_node_ids);
+    convert_gids_to_sonata(sonata_node_ids, population_offset_);
     hdf5_writer_->write(reports_population_group + "/mapping/node_ids", sonata_node_ids);
     hdf5_writer_->write(reports_population_group + "/mapping/index_pointers", index_pointers_);
     hdf5_writer_->write(reports_population_group + "/mapping/element_ids", element_ids_);
     hdf5_writer_->write_time(reports_population_group + "/mapping/time", time_);
     hdf5_writer_->configure_attribute(reports_population_group + "/mapping/time", "units", "ms");
 }
 
-void SonataData::write_spikes_header(const std::string& order_by) {
+void SonataData::write_spikes_header(const std::unique_ptr<Population>& population) {
     logger->trace("Writing SPIKE header!");
+    const std::string order_by = population->get_sorting();
     if (order_by != "by_time" && order_by != "by_id" && order_by != "none") {
         throw std::runtime_error("Order method " + order_by + "does not exists");
     }
 
-    const std::string spikes_population_group = "/spikes/" + population_name_;
+    const std::string spikes_population_group = "/spikes/" + population->get_population_name();
     hdf5_writer_->configure_group("/spikes");
     hdf5_writer_->configure_group(spikes_population_group);
     hdf5_writer_->configure_enum_attribute(spikes_population_group, "sorting", order_by);
-    hsize_t timestamps_size = Implementation::get_global_dims(report_name_,
-                                                              spike_timestamps_.size());
-    Implementation::sort_spikes(spike_timestamps_, spike_node_ids_, order_by);
-    hdf5_writer_->write(spikes_population_group + "/timestamps", spike_timestamps_);
+    hsize_t timestamps_size =
+        Implementation::get_global_dims(report_name_, population->get_spike_timestamps().size());
+    Implementation::sort_spikes(population->get_spike_timestamps(),
+                                population->get_spike_node_ids(),
+                                order_by);
+    hdf5_writer_->write(spikes_population_group + "/timestamps",
+                        population->get_spike_timestamps());
     if (timestamps_size > 0) {
         hdf5_writer_->configure_attribute(spikes_population_group + "/timestamps", "units", "ms");
     }
-    std::vector<uint64_t> sonata_spike_node_ids(spike_node_ids_);
-    convert_gids_to_sonata(sonata_spike_node_ids);
+    std::vector<uint64_t> sonata_spike_node_ids(population->get_spike_node_ids());
+    convert_gids_to_sonata(sonata_spike_node_ids, population->get_population_offset());
     hdf5_writer_->write(spikes_population_group + "/node_ids", sonata_spike_node_ids);
 }
 
+void SonataData::add_population(std::unique_ptr<Population>& population) {
```

Comment:
```suggestion
void SonataData::add_population(std::unique_ptr<Population> population) {
```
otherwise at the call site
```c++
auto foo = std::make_unique<Population>(...);
auto* foo_p = foo.get();
data.add_population(foo); // [surprisingly] leaves foo in moved-from state (foo_p still OK)
```
with the proposed change we would write
```c++
data.add_population(std::move(foo)); // explicit that foo is left in moved-from state
```

Reply:
Done